test/Microsoft.OpenApi.Tests/Validations/OpenApiSchemaValidationTests.cs

```diff
@@ -279,37 +279,18 @@ public void ValidateOneOfSchemaPropertyNameContainsPropertySpecifiedInTheDiscrim
                 {
                     {
                         "Person",
-                        new OpenApiSchema
-                        {
-                            Type = "array",
-                            Discriminator = new OpenApiDiscriminator
-                            {
-                                PropertyName = "type"
-                            },
-                            OneOf = new List<OpenApiSchema>
-                            {
-                                new OpenApiSchema
-                                {
-                                    Properties =
-                                    {
-                                        {
-                                            "type",
-                                            new OpenApiSchema
-                                            {
-                                                Type = "array"
-                                            }
-                                        }
-                                    },
-                                    Reference = new OpenApiReference
-                                    {
-                                        Type = ReferenceType.Schema,
-                                        Id = "Person"
-                                    }
-                                }
-                            },
-                            Reference = new OpenApiReference { Id = "Person" }
-                        }
-                    }                    
+                        new JsonSchemaBuilder()
+                            .Type(SchemaValueType.Array)
+                            .Discriminator("type", null, null)
+                            .OneOf(
+                                new JsonSchemaBuilder()
+                                    .Properties(
+                                        ("type", new JsonSchemaBuilder().Type(SchemaValueType.Array))
+                                    )
+                                    .Ref("#/components/schemas/Person")
+                            )
+                            .Ref("Person")
```

Comment:
I'm not sure what the difference between these refs is supposed to be.

The original code is

```c#
                                    Reference = new OpenApiReference	
                                    {	
                                        Type = ReferenceType.Schema,	
                                        Id = "Person"	
                                    }	
                                }	
                            },	
                            Reference = new OpenApiReference { Id = "Person" }
```

What does this reference do?  My `.Ref()` extension creates a `$ref` keyword, but I don't think that's really what's supposed to be happening here.

Reply:
Components have a self-reference.  It's an ugly implementation detail due to the way that we load $refs in as a dummy empty object (because the target of the reference hasn't been parsed yet) and then post-reading we run a reference resolver to replace the dummy objects with the target object.  This means the inline object and the component object are actually the same object instance.  Therefore, objects in the components collection also need to have a Reference property set or the writers cannot write out the $ref.

This model is going to change OpenAPI.Net v2 to a proxy pattern.  We need to do this because in OpenAPI v3.1 you can now have summary and description properties that sit along side the $ref.  With the current model we have no place to store that additional information.

Anyway, I think the proxy pattern for handling references is going to be a much better model anyway as it will not require doing the post-reading reference resolution step.  It also makes writing much easier.  You can see my attempt to implement the proxy pattern here https://github.com/microsoft/OpenAPI.NET/blob/dm/ProxyReference/src/Microsoft.OpenApi/Models/OpenApiSchemaReference.cs However, I did it on OpenAPISchema which is no longer appropriate since 3.1 made the distinction between OpenAPI reference objects and JSON Schema $ref.

Reply:
Okay, so I can just remove all of those, probably.  I'll need to go through and re-update all of those tests.  (I'll probably end up closing this and cherry-picking it to another branch for a different PR.)